### PR TITLE
Mac install

### PR DESCRIPTION
**Pull request by freesig**
This adds mac moltenvk automatic installation to Nannou.
Would love some feedback / a second pair of eyes.

https://github.com/mitchmindtree/vulkano/pull/4 needs to merge before this will work

To test comment out these in your `.bash_profile`
```
#export VULKAN_SDK="$HOME/.vulkan_sdk/macOS"

#export PATH="$VULKAN_SDK/bin:$PATH"

#export DYLD_LIBRARY_PATH="$VULKAN_SDK/lib:$DYLD_LIBRARY_PATH"

#export VK_ICD_FILENAMES="$VULKAN_SDK/etc/vulkan/icd.d/MoltenVK_icd.json"

#export VK_LAYER_PATH="$VULKAN_SDK/etc/vulkan/explicit_layer.d"
```
You also might want to
`cd ~`
 `mv .vulkan_sdk/ vulkan_sdk_BAK/ `
or if confident 
 `rm -fr .vulkan_sdk/`
Before the test.

Then just run any nannou application.

### CODE REVIEW

File: src/app.rs [hunk 1]
```diff
@@ -350,6 +352,7 @@ where
     /// The Model that is returned by the function is the same model that will be passed to the
     /// given event and view functions.
     pub fn new(model: ModelFn<M>) -> Self {
+        check_moltenvk();
```
[the remark this answers is not in the document]
Comment (freesig) [reply src/app.rs]: I think maybe that would be after `vulkano::instance(...)` is called though?

[the remark this answers is not in the document]
Comment (mitchmindtree) [reply src/app.rs]: I mean doing it around [here](https://github.com/nannou-org/nannou/blob/v0.9/src/app.rs#L514). The vulkan instance doesn't get created until a few lines below there.

However, if the user wants to create their own instance and pass it into nannou via the `.vulkan_instance(instance)` method, than they'll need to run `moltenvk_deps` at some point themselves before they building the `app` anyway, and this should be such a rare case that I think it's fine to expect?

**Comment by freesig** [comment]
I like the idea of it being part of the builder method. And I think it's fine if the default just runs silently. The hard part is adding a builder method that is macos only.

> Also, whenever you see a `(Y/n)`

I thought I fixed this, is enter not working?

> perhaps the `check_or_install` function could have some way of accepting a `prompt` function

I think this could work but you would really need a `start` `progress` and `complete` callback. 

> Progress Indicator

I could perhaps make this part of the `default()` if they don't specify callbacks. 

## Concerns
I guess my main concern about this being completely silent is that it means every nannou application on mac becomes pretty intrusive. Like if someone sends you there `.bin` and is like hey check out my cool sketch, then suddenly it's downloading moltenvk and editing your `.bash_profile` and you didn't even know there was a readme.

I'm happy to make a silent option but I think it should be opt in not opt out. I'm happy for the default prompt to not even be a question but I think it should at least print a message explaining what is happening.

I just think I would be pretty annoyed if someone sent me a binary and it did all that without saying what was happening.

**Comment by freesig** [comment]
also any idea why this travis build failed?

File: src/app.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (freesig) [reply src/app.rs]: Yeh that should be fine because if they have already run `moltenvk_deps` then it will just continue silently anyway

**Comment by mitchmindtree** [comment]
> I thought I fixed this, is enter not working?

Ahh josh just said that he tried  using lowercase `y` and brought it up. I *think* `y` of any case or enter should work, but honestly it probably doesn't matter haha.

> I think this could work but you would really need a `start` `progress` and `complete` callback.

Yes that would be even better, perhaps it could be a `event` callback like `FnMut(Event) -> bool` where `Event` is something like `pub enum Event { Start, Progress(f32), Complete }` - I'm not sure what the `bool` would mean in response to `Complete` though, so maybe your suggestion of using different functions is better.

> I could perhaps make this part of the `default()` if they don't specify callbacks.

Yeah I think that sounds reasonable? I wouldn't mind if it didn't do anything by default, as long as there's some sort of callback, e.g. we could always visualise it in the terminal ourselves in nannou by default using the callback, even if that's not the default behaviour in moltenvk_deps. Either way!

> Concerns

This is a really good point! I hadn't thought of it from the sketch-sharing perspective. I guess one thing that would make it nicer would be to add a comment over the lines added to the `.bash_profile` saying something like `# Added by nannou 2019.02.14 for moltenvk installation.`? That said, I totally agree that making the user-prompt an opt-out thing might be more reasonable - I'd be happy to go with that if you'd prefer!

> also any idea why this travis build failed?

Ahh I think you just need to rebase onto v0.9 again, looks like you're missing #268.

**Comment by freesig** [comment]
I'm a bit stuck on this progress bar function because curl wants a
`FnMut(f64, f64, f64, f64) -> bool + Send + 'static`
But I can't get a callback with a `'static` lifetime.
I've tried using a lazy_static! but that only gives you a `Mutex<Option<FnMut(f64, f64, f64, f64) -> bool + Send>` without the static lifetime.

I think the best I can do is give the use a standard fn pointer as a callback. Is this too restrictive? TBH this is starting to seem a bit beyond the scope of a crate that is just meant to download the dependencies.

**Comment by freesig** [comment]
I've made pretty much all the changes see https://github.com/nannou-org/nannou/issues/233#issuecomment-463116666
The last this to do is update the App builder so that nannou users can alter how the install works.

**Comment by freesig** [comment]
This should be all good to go now. Not completely sure on the Builder methods. Maybe the doc comment could be clearer

File: src/gpu.rs [hunk 2]
```diff
@@ -250,3 +262,53 @@ pub fn msaa_samples_limited(physical_device: &PhysicalDevice, target_msaa_sample
...
+        khr_surface: true,
+        khr_xlib_surface: true,
+        khr_xcb_surface: true,
+        khr_wayland_surface: true,
+        khr_android_surface: true,
+        khr_win32_surface: true,
+        mvk_ios_surface: true,
+        mvk_macos_surface: true,
+        ..InstanceExtensions::none()
+    };
```
Comment (mitchmindtree) [inline src/gpu.rs]: This looks like the same extensions as required by `vulkano-win` [here](https://github.com/vulkano-rs/vulkano/blob/master/vulkano-win/src/lib.rs#L30) - should we be using `vulkano_win::required_extensions()` instead? Where did you get this list from?

I'm not 100% sure I understand this function - it says `required_extensions_with_loader` but then it returns the intersection of ideal and supported extensions?

Comment (mitchmindtree) [reply src/gpu.rs]: Ahhh I see, this is a copy of `vulkano_win::required_extensions` but your checking what's supported by core using the custom loader?

Comment (freesig) [reply src/gpu.rs]: It is indeed the same function except that you can give it a loader. The one in `vulkano_win` uses `auto_loader` which will not work when the env variables are not sourced. So in order to allow installation to work without requiring the user to source `.bash_profile` I need to provide a loader with a specific path to the lib. This loader also needs to be user to get the `required_extensions` unfortunately `vulkano_win` does not provide this ability.

Another odd fact is that everything in that function is actually just in `vulkano` so I think it's a little odd that it is even in `vulkano-win`

**Comment by mitchmindtree** [comment]
This looks good to me!

**Pull request merged**